--- a/pythonforandroid/bootstraps/webview/build/src/main/java/org/kivy/android/PythonActivity.java
+++ b/pythonforandroid/bootstraps/webview/build/src/main/java/org/kivy/android/PythonActivity.java
@@ -34,6 +34,7 @@ import android.view.ViewGroup.LayoutParams;
 
 import android.webkit.WebViewClient;
 import android.webkit.WebView;
+import android.webkit.CookieManager;
 import android.net.Uri;
 
 import org.renpy.android.ResourceManager;
@@ -165,16 +166,21 @@ public class PythonActivity extends Activity {
             mWebView.setWebViewClient(new WebViewClient() {
                     @Override
                     public boolean shouldOverrideUrlLoading(WebView view, String url) {
+                        Uri u = Uri.parse(url);
                         if (mOpenExternalLinksInBrowser) {
-                            if (!(url.startsWith("file:") || url.startsWith("http://127.0.0.1:"))) {
-                                Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
+                            if (!(u.getScheme().equals("file") || u.getHost().equals("127.0.0.1"))) {
+                                Intent i = new Intent(Intent.ACTION_VIEW, u);
                                 startActivity(i);
                                 return true;
                             }
                         }
-                        view.loadUrl(url);
                         return false;
                     }
+
+                    @Override
+                    public void onPageFinished(WebView view, String url) {
+                        CookieManager.getInstance().flush();
+                    }
                 });
             mLayout = new AbsoluteLayout(PythonActivity.mActivity);
             mLayout.addView(mWebView);
