--- a/pythonforandroid/bootstraps/common/build/src/main/java/org/kivy/android/PythonUtil.java
+++ b/pythonforandroid/bootstraps/common/build/src/main/java/org/kivy/android/PythonUtil.java
@@ -44,6 +44,7 @@ public class PythonUtil {
         libsList.add("python3.6m");
         libsList.add("python3.7m");
         libsList.add("python3.8");
+        libsList.add("python3.9");
         libsList.add("main");
         return libsList;
     }
@@ -63,7 +64,7 @@ public class PythonUtil {
                 // load, and it has failed, give a more
                 // general error
                 Log.v(TAG, "Library loading error: " + e.getMessage());
-                if (lib.startsWith("python3.8") && !foundPython) {
+                if (lib.startsWith("python3.9") && !foundPython) {
                     throw new RuntimeException("Could not load any libpythonXXX.so");
                 } else if (lib.startsWith("python")) {
                     continue;
